src/NLPModelsIpopt.jl

```diff
@@ -76,22 +114,60 @@ function ipopt(nlp::AbstractNLPModel; callback::Union{Function, Nothing} = nothi
     end
   end
 
-  problem = CreateIpoptProblem(
-    n,
-    nlp.meta.lvar,
-    nlp.meta.uvar,
-    m,
-    nlp.meta.lcon,
-    nlp.meta.ucon,
-    nlp.meta.nnzj,
-    nlp.meta.nnzh,
-    eval_f,
-    eval_g,
-    eval_grad_f,
-    eval_jac_g,
-    eval_h,
-  )
+  return eval_f, eval_g, eval_grad_f, eval_jac_g, eval_h
+end
 
+"""
+    output = ipopt(nlp; kwargs...)
+
+Solves the `NLPModel` problem `nlp` using `IpOpt`.
+
+For advanced usage, first define a `IpoptSolver` to preallocate the memory used in the algorithm, and then call `solve!`:
+    solver = IpoptSolver(nlp)
+    solve!(solver, nlp; kwargs...)
+    solve!(solver, nlp, stats; kwargs...)
+
+# Optional keyword arguments
+* `x0`: a vector of size `nlp.meta.nvar` to specify an initial primal guess
+* `y0`: a vector of size `nlp.meta.ncon` to specify an initial dual guess for the general constraints
+* `zL`: a vector of size `nlp.meta.nvar` to specify initial multipliers for the lower bound constraints
+* `zU`: a vector of size `nlp.meta.nvar` to specify initial multipliers for the upper bound constraints
+
+All other keyword arguments will be passed to IpOpt as an option.
+See https://coin-or.github.io/Ipopt/OPTIONS.html for the list of options accepted.
+
+# Output
+The returned value is a `GenericExecutionStats`, see `SolverCore.jl`.
+
+# Examples
+```
+using NLPModelsIpopt, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3));
+stats = ipopt(nlp, print_level = 0)
+```
+
+```
+using NLPModelsIpopt, ADNLPModels
+nlp = ADNLPModel(x -> sum(x.^2), ones(3));
+solver = IpoptSolver(nlp);
+stats = solve!(solver, nlp, print_level = 0)
+```
+"""
+function ipopt(nlp::AbstractNLPModel; kwargs...)
+  solver = IpoptSolver(nlp)
+  stats = GenericExecutionStats(nlp)
+  return solve!(solver, nlp, stats; kwargs...)
+end
+
+function SolverCore.solve!(
+  solver::IpoptSolver,
+  nlp::AbstractNLPModel,
+  stats::GenericExecutionStats;
+  callback::Union{Function, Nothing} = nothing,
```

Comment:
You will not help the compiler with `Union{Function, Nothing}`. You should not give a type to `callback` because the compiler is unable to infer the output of a function and add `::nothing` where we call the callback such that the compiler knows the output type of the `callback`.
https://github.com/JuliaSmoothOptimizers/Krylov.jl/blob/main/src/cg.jl#L203  

If you also restrict callback::Function, you won't be able to use a "functor".
https://docs.julialang.org/en/v1/manual/methods/#Function-like-objects

Reply:
I see. I opened a PR #95 to fix this, and then I will rebase.